### PR TITLE
mgr: enable rook orchestrator mgr module by default

### PR DESCRIPTION
**Pull request by rkachach**
previously, we had kept the Rook orchestrator manager module disabled so far because it was causing a bunch of issues and errors on the dashboard. But with the recent changes made in the `cephv v18.2.1` release, we've fixed those issues and made some overall improvements to the dashboard user experience when Rook is enabled. With all that in mind, it's time to switch on the Rook orchestrator by default.

Fixes: https://github.com/rook/rook/issues/13760

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: deploy/examples/cluster.yaml [hunk 1]
```diff
@@ -59,8 +59,8 @@ spec:
     modules:
       # List of modules to optionally enable or disable.
       # Note the "dashboard" and "monitoring" modules are already configured by other settings in the cluster CR.
-      # - name: rook
-      #   enabled: true
+      - name: rook
```
Comment (parth-gr) [inline deploy/examples/cluster.yaml:62]: cluster-on-pvc.yaml too

File: deploy/examples/cluster-test.yaml [hunk 2]
```diff
@@ -23,6 +23,9 @@ spec:
   mgr:
     count: 1
     allowMultiplePerNode: true
+    modules:
```
Comment (parth-gr) [inline deploy/examples/cluster-test.yaml:26]: can you summarize a little bit I always confuse this with dashboard mgr

**Review by parth-gr — CHANGES_REQUESTED** [review]

File: deploy/examples/cluster-test.yaml (hunk 2 above)
Comment (rkachach) [reply deploy/examples/cluster-test.yaml:26]: @parth-gr dashobard is a mgr module that calls the orchestrator -- another mgr module-- (rook or cephadm) to get some information such as daemons, devices, etc. Does this answer your question?

**Review by subhamkrai — COMMENTED** [review]
A thought, can check in go code if dashboard is enabled then only enable mgr module, or there are other uses cases of mgr module? 

IIRC in the past it was true only but we changed that to false since we have some issue probably in dashboard or somewhere. Those error may not exist today but just sharing the thoughts.

File: deploy/examples/cluster-test.yaml (hunk 2 above)
Comment (parth-gr) [reply deploy/examples/cluster-test.yaml:26]: so to use dashboard we need to always turn it on?

**Comment by travisn** [comment]
> A thought, can check in go code if dashboard is enabled then only enable mgr module, or there are other uses cases of mgr module?
> 
> IIRC in the past it was true only but we changed that to false since we have some issue probably in dashboard or somewhere. Those error may not exist today but just sharing the thoughts.

Agreed, if the dashboard module is enabled, it will be good to automatically enable the rook module. Perhaps we don't even need to support enabling/disabling the `rook` module if it's listed in the `modules` section of the cluster CR. So this likely means when the `configureDashboardModules()` method is called to enable the dashboard, that method would call `configureOrchestratorModules()`.

File: deploy/examples/cluster-test.yaml (hunk 2 above)
Comment (rkachach) [reply deploy/examples/cluster-test.yaml:26]: @parth-gr not really, some parts of the dashboard don't depend on the orch but others do. So dashboard partially depend on the orchestrator.

**Comment by travisn** [comment]
> Still not clear what are the benefits of doing this in code as we have the support to enable/disable modules through configuration. I'd prefer to keep this kind of functionality in conf unless there's really a good reason to move it to code.

The main question for me is if we want to consider it's "required" to have the `rook` mgr module enabled when the dashboard is enabled. When it's not enabled, the dashboard will show messages about needing to enable it, so it really seems like it's required for full dashboard functionality. 

In practice, it will likely be sufficient to just add the `rook` mgr module to the example manifests for enabling it as a separate module, so implementing it in code won't really change the scenario.

**Comment by travisn** [comment]
Since v18.2.2 is out, it would be great to enable this by default now. How about just rebasing to get the latest CI fixes?

**Comment by rkachach** [comment]
> Since v18.2.2 is out, it would be great to enable this by default now. How about just rebasing to get the latest CI fixes?

Done

**Review by travisn — COMMENTED** [review]
Looks like a permission is missing for tests where there is a cluster in a second namespace:
```
2024-03-13 19:29:22.737236 I | ceph-spec: ceph-file-controller: CephCluster "my-cluster" found but skipping reconcile since ceph health is &{Health:HEALTH_ERR Details:map[MGR_MODULE_ERROR:{Severity:HEALTH_ERR Message:Module 'rook' has failed: (403)
Reason: Forbidden
HTTP response headers: HTTPHeaderDict({'Audit-Id': '6b518229-d6e1-49d1-a16b-247e2eab237f', 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'X-Content-Type-Options': 'nosniff', 'X-Kubernetes-Pf-Flowschema-Uid': '2609213b-bd67-4367-9804-46c4e37b0667', 'X-Kubernetes-Pf-Prioritylevel-Uid': 'f640617f-7d82-45fd-a24b-f5ffd9e34525', 'Date': 'Wed, 13 Mar 2024 19:19:40 GMT', 'Content-Length': '295'})
HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"nodes is forbidden: User \"system:serviceaccount:rook-ceph-secondary:rook-ceph-mgr\" cannot list resource \"nodes\" in API group \"\" at the cluster scope","reason":"Forbidden","details":{"kind":"nodes"},"code":403}

}] LastChecked:2024-03-13T19:28:50Z LastChanged:2024-03-13T19:20:42Z PreviousHealth:HEALTH_OK Capacity:{TotalBytes:6442450944 UsedBytes:28209152 AvailableBytes:6414241792 LastUpdated:2024-03-13T19:28:50Z} Versions:0xc00156fc00 FSID:3be4c34b-e180-4f6b-a346-0034aa37a77d}
```

**Comment by rkachach** [comment]
The full stack trace from the error: 

```
Reason: Forbidden
HTTP response headers: HTTPHeaderDict({'Audit-Id': 'c70097f3-3d57-4f6b-b7ad-b10435fcda43', 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'X-Content-Type-Options':
 'nosniff', 'X-Kubernetes-Pf-Flowschema-Uid': 'd3b30d1e-d8d1-45ce-9279-fd7cf81515a9', 'X-Kubernetes-Pf-Prioritylevel-Uid': '84562724-007b-4670-afc5-1a77fb588000', 'Date': 'Thu, 14 Mar 2024 1
5:36:37 GMT', 'Content-Length': '295'})
HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"nodes is forbidden: User \"system:serviceaccount:rook-ceph-secondary:rook-ceph-mgr\" cannot
 list resource \"nodes\" in API group \"\" at the cluster scope","reason":"Forbidden","details":{"kind":"nodes"},"code":403}

Traceback (most recent call last):
  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 137, in wrapper
    return OrchResult(f(*args, **kwargs))
  File "/usr/share/ceph/mgr/rook/module.py", line 244, in get_hosts
    return self.rook_cluster.get_hosts()
  File "/usr/share/ceph/mgr/rook/rook_cluster.py", line 1230, in get_hosts
    for node in self.nodes.items:
  File "/usr/share/ceph/mgr/rook/rook_cluster.py", line 302, in items
    resource_version = self._fetch()
  File "/usr/share/ceph/mgr/rook/rook_cluster.py", line 284, in _fetch
    response = self.api_func(**self.kwargs)
  File "/usr/lib/python3.6/site-packages/kubernetes/client/api/core_v1_api.py", line 13726, in list_node
    (data) = self.list_node_with_http_info(**kwargs)  # noqa: E501
  File "/usr/lib/python3.6/site-packages/kubernetes/client/api/core_v1_api.py", line 13821, in list_node_with_http_info
    collection_formats=collection_formats)
  File "/usr/lib/python3.6/site-packages/kubernetes/client/api_client.py", line 345, in call_api
    _preload_content, _request_timeout)
  File "/usr/lib/python3.6/site-packages/kubernetes/client/api_client.py", line 176, in __call_api
    _request_timeout=_request_timeout)
  File "/usr/lib/python3.6/site-packages/kubernetes/client/api_client.py", line 366, in request
    headers=headers)
  File "/usr/lib/python3.6/site-packages/kubernetes/client/rest.py", line 241, in GET
    query_params=query_params)
  File "/usr/lib/python3.6/site-packages/kubernetes/client/rest.py", line 231, in request
    raise ApiException(http_resp=r)
kubernetes.client.rest.ApiException: (403)
Reason: Forbidden

```

File: deploy/examples/common-second-cluster.yaml [hunk 3]
```diff
@@ -288,3 +288,57 @@ kind: ServiceAccount
...
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: rook-ceph-mgr-secondary-cluster
+subjects:
+  - kind: ServiceAccount
+    name: rook-ceph-mgr
+    namespace: rook-ceph # namespace:cluster
+---
+# Aspects of ceph-mgr that require cluster-wide access
+kind: ClusterRole
```
Comment (travisn) [inline deploy/examples/common-second-cluster.yaml:307]: Since this clusterrole is already defined in common.yaml, we should actually be able to assume it already is defined. ClusterRoles only need to be defined once, if there is no difference needed in the different namespace. No need for defining this ClusterRole here.

File: deploy/examples/common-second-cluster.yaml [hunk 4]
```diff
@@ -288,3 +288,57 @@ kind: ServiceAccount
...
+metadata:
+  name: rook-ceph-mgr-secondary-cluster
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: rook-ceph-mgr-secondary-cluster
+subjects:
+  - kind: ServiceAccount
+    name: rook-ceph-mgr
+    namespace: rook-ceph # namespace:cluster
```
Comment (travisn) [inline deploy/examples/common-second-cluster.yaml:304]: I believe this namespace fix on the binding will be sufficient.
```suggestion
    namespace: rook-ceph-secondary # namespace:cluster
```

**Review by travisn — CHANGES_REQUESTED** [review]

File: deploy/examples/common-second-cluster.yaml (hunk 4 above)
Comment (rkachach) [reply deploy/examples/common-second-cluster.yaml]: That fixed it + adjusting the name. Thanks @travisn.

**Review by travisn — APPROVED** [review]

File: deploy/examples/common-second-cluster.yaml [hunk 5]
```diff
@@ -288,3 +288,17 @@ kind: ServiceAccount
 metadata:
   name: rook-ceph-rgw
   namespace: rook-ceph-secondary # namespace:cluster
+---
+# Allow the ceph mgr to access cluster-wide resources necessary for the mgr modules
```
[the remark this answers is not in the document]
Comment (parth-gr) [reply deploy/examples/common-second-cluster.yaml:292]: np I see a separate commit for this

[the remark this answers is not in the document]
Comment (travisn) [reply deploy/examples/common-second-cluster.yaml:292]: The rook mgr module requires these permissions
